### DISCUSSION TITLE
Hide columns by default in V8

### DISCUSSION THREAD

**Discussion by Schmell** [TanStack/table#4106] [Q&A]
I read this [discussion](https://github.com/TanStack/table/discussions/1971) abut hiding columns in v7
The docs suggest that there is a initialState.hiddenColumns in the use(Framework)Table , but this is throwing, saying that hiddenColumns does not exist
Maybe i am putting it in the wrong spot or not writing it right

**Comment by KevinVandy** [comment]
initialState.columnVisibility in V8

**Reply by Schmell** [reply]
Sorry, but i am still un-clear what i pass to this
short example please
also do i need to enableHiding in colDef

**Reply by KevinVandy** [reply]
I recommend reading the columnVisibility api docs https://tanstack.com/table/v8/docs/api/features/column-visibility

You can see the exact typescript types that they are expecting.

Here's an example of what you would pass to useReactTable options.

```jsx
initialState: { columnVisibility: { firstName: false } }
```

**Reply by Schmell** [reply]
Thanks for your reply @KevinVandy .
I have read the docs and tried that exact approach.
In useReactTable I write -
`data,
columns,
`
```javascript
initialState: {
      columnVisibility: {
        boat: true,
        points: true,
        corrected: true,
        elapsed: false,
        finish: false,
        nett: true,
      },
      pagination: {
        pageSize: 30,
      },
      columnPinning: {
        right: ["nett"],
      },
    },
```

But this does not work. my "elapsed" and "finish" columns still render regardless of enableHiding

Thanks again for your help.

**Reply by tannerlinsley** [reply]
You need to use the column visibility state aware methods of the table. Take a look at the table apis used in the column visibility example.

**Reply by cereallarceny** [reply]
@tannerlinsley I'm getting the same result as @Schmell. You're saying that we need to use the column visibility methods that the table gives us to initially set this on table initialization... but then what's the point of passing initialState at all? Is this a vestige of the old way of doing things? The docs don't make this terribly clear.

**Reply by cereallarceny** [reply]
I should add, in the same way that I can pass a `size` key to the column definition, I'm a bit surprised something like a `hidden` key doesn't exist. I've had to augment the internal Typescript classes to support this functionality. Any pointers in the right direction here?

**Reply by KevinVandy** [reply]
Can you follow the [column visibility example](https://tanstack.com/table/v8/docs/examples/react/column-visibility) from the docs?

**Reply by Schmell** [reply]
I believe i solved this by putting the values in state: not initialState:

```javascript
  const table = useReactTable({
    data,
    columns,
    state: {
      sorting,
      columnVisibility, // This an object of columns with true or false
      columnPinning,
    },
    initialState: {
      pagination: {
        pageSize: 30,
      },
    },
...
```

**Reply by KP1104** [reply]
Hi i am using v8 react table and i want to hide 2 columns from my 10 columns. How should i do it?

**Reply by wiskani** [reply]
Hi just make this: 
Make a useState, with inicial hidde columns (id and bolean)
`
const [columnVisibility, setColumnVisibility] = React.useState({id:false})
`
Then put columnVisibility in state :
`
const table = useReactTable({
    data,
    columns,
    state: {
      columnVisibility,
    },
`
After you can change state of columns visibilty of those you want

**Reply by KP1104** [reply]
Ok. Is there a way if i can pass it to column def directly? Like show or enable hiding true?

**Reply by wiskani** [reply]
I don't know about that, but in the initial state of columnVisibility you can set the visibility of each column using its id. In version 7 it could be done in the definition of the columns but here it seems better to me since later you can change the visibility state more easily if you need it

**Reply by kndshein** [reply]
Background: I'm pretty new to TS, and very new to this library.
Somehow `initialState.columnVisibility` didn't do jack. My first inclination was to try what @Schmell did, but I hit a type error:
```TS
 Type 'Dispatch<SetStateAction<{ is_verified: boolean; is_disabled: boolean; }>>' is not assignable to type 'OnChangeFn<VisibilityState>'.
```

By setting `<VisibilityState>` on my `useState`, the type lords became happy, and it worked! 
Rendering `{JSON.stringify(table.getState().columnVisibility, null, 2)` like in the example is quite helpful.

**Reply by KevinVandy** [reply]
Yep, if you manage column visibility in both state and initialState, state overwrites.

**Reply by KevinVandy** [reply]
^ Which means that you might just have to initialize the columnVisibility `state`, instead of using the `initialState` prop. Doesn't mean that you can't still have columns initially hidden if managing the full state.

**Reply by maik-emanoel** [reply]
> Ok. Is there a way if i can pass it to column def directly? Like show or enable hiding true?

I wanted to do the same thing, did you find some way to do that?

**Comment by timurbesli** [comment]
On V8, You can use setColumnVisibility to change the visibility state like this.
`
setColumnVisibility({
    "headerId": true,
    "name": true,
    "price": false,
  })`
I created a saved hide/show effect on typescript. hope its helpful.

```
const [columnVisibility, setColumnVisibility] = React.useState<VisibilityState>({});

const table = useReactTable({
    columns: columns,
    data: data,
    state: {
      columnVisibility: columnVisibility,
    },

    getCoreRowModel: getCoreRowModel(),
    onColumnVisibilityChange: setColumnVisibility,
  });
  

useEffect(() => {
    if (localStorage.getItem("visibleColumnIds") === null) {
      const columIds = table.getAllColumns().map((column) => column.id);
      localStorage.setItem("visibleColumnIds", JSON.stringify(columIds));
    }
    const allHideableColumnIds = table
      .getAllColumns()
      .filter((column) => column.getCanHide())
      .map((column) => column.id);
    const savedColumnIds: string[] = JSON.parse(localStorage.getItem("visibleColumnIds") || "[]") as string[];
    const initialColumnVisibility: VisibilityState = allHideableColumnIds.reduce((visibilityState, columnId) => {
      visibilityState[columnId] = savedColumnIds.includes(columnId);
      return visibilityState;
    }, {} as VisibilityState);

    setColumnVisibility(initialColumnVisibility);
  }, []);
```

**Reply by Ericokim** [reply]
Hi @timurbesli following your example above, I made some tweaks see the code below;
```js

  const [columnVisibility, setColumnVisibility] = React.useState({});

  const columns = React.useMemo(
    () => [
      {
        accessorKey: 'firstName',
        enableHiding: false,
        header: 'First Name',
      },
      {
        accessorKey: 'lastName',
        enableHiding: false,
        header: 'Last Name',
      },
      {
        accessorKey: 'address',
        enableHiding: true,
        show: false,     // Use this state to hide/show your column `show: true || false`
        header: 'Address',
      `},`
    ],
    [],
  );

//  This will check if all the columns has `show: false` .
  React.useEffect(() => {
    const columnIds = columns.map((col) => col.id);
    const initialColumnVisibility = columns
      .filter((col) => col.show === false)
      .map((col) => col.id)
      .reduce((visibilityState, columnId) => {
        visibilityState[columnId] = !columnIds.includes(columnId);
        return visibilityState;
      }, {});

    setColumnVisibility(initialColumnVisibility);
  }, []);

```

**Reply by LucasJantschChitolina** [reply]
This was very helpful, thanks a lot!

**Comment by 7sne** [comment]
In the `v8` it can be easily done just by providing the default values for the `columnVisibility` state.

```tsx
    const [columnVisibility, setColumnVisibility] = React.useState<VisibilityState>({
        'column-name-to-hide-by-default': false,
    })

   // (...)

    const table = useReactTable({
        // (...)
        state: {
            sorting,
            columnVisibility, // <-- State.
            columnFilters,
        },
       onColumnVisibilityChange: setColumnVisibility, // <-- To manipulate visibility state.
    })
```

**Comment by jschuur** [comment]
> In the `v8` it can be easily done just by providing the default values for the `columnVisibility` state.

Is there another way to pass in multiple data properties for a row that are rendered inside a inside single cell without defining them as columns that are always hidden?

Concrete example: I have a site `name` and `url` value. My `name` column shows the name with a hyperlink to the URL. I never want to list the `url` property as its own column. I just want to pass it in with my table `data` and reference later in a `row.getValue('url')`.

The only way I don't get a `app-index.js:31 [Table] Column with id 'url' does not exist.` error is if I add this to my `ColumnDef` and add a `VisiblityState` that includes `url: false`:

```js
{
  accessorKey: 'url',
},
```

It works, it just feels like I am abusing the concept of a 'column' that I never intended on displaying as such.

See also https://github.com/TanStack/table/discussions/4893

**Reply by alangshur** [reply]
Was also trying to figure this out and the solution for me was to use `row.original.url` instead of `row.getValue('url')` to reference the original data. Hope that helps!

**Reply by ahaseebcontractken** [reply]
This worked like magic! Thanks.

**Reply by benakdev** [reply]
> Was also trying to figure this out and the solution for me was to use `row.original.url` instead of `row.getValue('url')` to reference the original data. Hope that helps!

Best solution by far! Thanks!

**Comment by ZeldOcarina** [comment]
It's even simpler by using an effect and just hiding the columns you want to be hidden:
```js
React.useEffect(() => {
    table.setColumnVisibility({
      locationId: false,
      leadType: false,
      source: false,
    });
  }, [table]);
```

**Comment by navid-asgharzadeh** [comment]
```js
  const table = useReactTable({
        data,
        columns,
        state: {
            columnVisibility: {
                id: false //replace `id` with your column identifier
            }
        }
    });
```

**Comment by RaviRamDhali** [comment]
Using Ver 8: 

https://github.com/TanStack/table/discussions/1971#discussioncomment-8125281

Function GetInVisibleColumn() takes an array of finalColumnDef and filters out the columns that have the visible property set to false. It then creates a new object (removedColumn) where the keys are the accessorKey properties of the invisible columns, and the values are set to false.

**Comment by nivethan-me** [comment]
Use initialState.columnVisibility in v8 if you want to reset the columns 

```
  const table = useReactTable({
    data: data,
    columns: columns,
    enableColumnResizing: true,
    enableHiding: true,
    // set false to non default columns here so reset button hides them
    initialState: { columnVisibility: { customer: false } },
    columnResizeMode: "onChange",
    getCoreRowModel: getCoreRowModel(),
    debugTable: true,
    debugHeaders: true,
    debugColumns: true,
  });
  ```
  
  
  ```
            {table.getAllLeafColumns().map((column) => {
            return (
              <div key={column.id} className="px-1">
                <label>
                  <input
                    {...{
                      type: "checkbox",
                      checked: column.getIsVisible(),
                      onChange: column.getToggleVisibilityHandler(),
                    }}
                  />{" "}
                  {column.id}
                </label>
              </div>
            );
          })}
        </div>
        <button onClick={() => table.resetColumnVisibility()}>reset</button>
```

**Comment by salihayazdir** [comment]
If you want to manage the initial visibility in column definition:

Extend the column meta to add hidden key
```
declare module '@tanstack/react-table' {
  interface ColumnMeta<TData extends RowData, TValue> {
    hidden?: boolean;
  }
}
```
```
{
    accessorKey: 'materialName',
    accessorFn: (row) => row.material.name,
    meta: {
      hidden: true,
    },
...
  },

```

Add the columns you marked with hidden in the initial column visibility state
  ```
const table = useReactTable({
    data,
    columns,
    ...
    onColumnVisibilityChange: setColumnVisibility,
    state: {
      columnVisibility,
    },
  });
```
```
 const initialVisibility = columns.reduce((acc, column) => {
    if (column.meta?.hidden && (column as any)?.accessorKey) {
      acc[(column as any).accessorKey as string] = false;
    }
    return acc;
  }, {} as VisibilityState);

  const [columnVisibility, setColumnVisibility] = useState<VisibilityState>(initialVisibility);
```

I had to use any for the column definition because typescript didnt want me to use accessorKey. I dont know why.

**Reply by thierryntoh24** [reply]
This works perfectly. Thanks
PS. Make sure to change `accessorKey` to `id` if your column data is a nested object like in my case, where I have `accessorKey = 'city.address', id='address'`

**Reply by thierryntoh24** [reply]
I even went ahead and modified your solution a bit for initial sorting state.

```
declare module "@tanstack/react-table" {
  interface ColumnMeta<TData extends RowData, TValue> {
    hidden?: boolean;
    sort?: boolean; //desc or asc, true or false
  }
}
```

```
{
    ...
    meta: { hidden: true, sort: true },
}
```

```
  const initialSort = columns.reduce(
    (acc, column) => {
      if (column.meta?.sort && column?.id) {
        acc.push({ id: column.id, desc: column.meta?.sort });
      }
      return acc;
    },
    [{}] as SortingState,
  );
  const [sorting, setSorting] = React.useState<SortingState>(initialSort);
  ```

**Comment by riteshreg** [comment]
This works for me

` const [rowSelection, setRowSelection] = React.useState({});

  const [columnVisibility, setColumnVisibility] =
    React.useState<VisibilityState>({
      CompanyName: false,
    });
    
  const [columnFilters, setColumnFilters] = React.useState<ColumnFiltersState>(
    [],
  );
  const [sorting, setSorting] = React.useState<SortingState>([]);

  const [filtering, setFiltering] = React.useState<string>("");

  const table = useReactTable({
    data,
    columns,
    state: {
      globalFilter: filtering,
      sorting,
      columnVisibility,
      rowSelection,
      columnFilters,
    },`

**Comment by SaadFarooq-Dev** [comment]
If you are using a component to encapsulate the table logic, you can pass a prop called `columnVisibility` to manage the visibility of columns dynamically. In the `DataTable` component, you can set it up like this:

```ts
type prop  = {
  columnVisibility?: Record<string, boolean>;
}
  ```

```ts
// In your DataTable component
state: {
  columnVisibility: columnVisibility,
},
```

Parent Component. 
``` ts
        <DataTable
          data={logs}
          isLoading={isLoading}
          columnVisibility={{
            updated: false // You can also do this conditionally based on a state. 
          }} 
          />
```
